rfcs/20190411-tensorboard-improved-plugin-ext.md

```diff
@@ -0,0 +1,151 @@
+
+In order to be discovered by TensorBoard, the distribution should include an [`entry_points` entry][entry-points] declaring itself as a TensorBoard plugin. The entry point group is `tensorboard_plugins`. The entry point name is arbitrary, and ignored. The entry point object reference should refer to the standard interface described in the previous paragraph. As an example, the setup.py file for a hypothetical “widgets” plugin might look like:
+
+```python
+import setuptools
+setuptools.setup(
+    name="tensorboard_plugin_widgets",
+    description="Widget chart support for TensorBoard",
+    packages=setuptools.find_packages(),
+    # more metadata (version, license, authors…)
+    entry_points={
+        "tensorboard_plugins": [
+            "widgets = tensorboard_plugin_widgets.plugin:get_plugin",
+        ],
+    },
+)
+```
+
+assuming that the `tensorboard_plugin_widgets.plugin` module exposes a function `get_plugin` conforming to the interface described above.
+
+To facilitate plugin development without requiring plugins to be packaged into a wheel and installed into a virtualenv, TensorBoard will learn an `--extra_plugins` flag, whose value will be a list of comma-separated object references. For instance, for testing the widgets plugin, one might invoke
+
+```
+tensorboard --logdir /tmp/whatever \
+    --extra_plugins tensorboard_plugin_widgets.plugin:get_plugin
+```
+
+with the `tensorboard_plugin_widgets.plugin` module on the Python path.
+
+[entry-points]: https://packaging.python.org/specifications/entry-points/
+[tb-loader]: https://github.com/tensorflow/tensorboard/blob/dc71eeea403eb027612422ab6cd3061d4f9a10f3/tensorboard/plugins/base_plugin.py#L149-L162
+
+### Frontend binary loading
+One of our soft goals is to enable existing visualization suite to integrate with TensorBoard with ease without having to change, for instance, module system. Of course, a plugin has to procure data by writing a backend and some glue code to integrate with our backend, but we are hoping that most of the visualization part can remain untouched.
+
+#### Requirement: Colab Integration
+TensorBoard recently added support for Jupyter and Colab and the feature is well received by the community. The new plugin frontend design should be compatible with Colab integration.
+
+One of the critical limitation Colab imposes TensorBoard is the usage of iframe and it is related to how it routes requests to the backend. To make it transparent to TensorBoard, Colab uses a ServiceWorker to intercept and reroute requests to localhost via its internal API. However, due to the implementation of Colab's ServiceWorker that is designed to intercept only requests to localhost (cannot use a relative path), one cannot render an iframe in an output cell.
+
+This restriction is relevant for *all solutions* below since a plugin can decide to render an iframe that makes a request anywhere.
+
+#### Non-iframe
```

Comment:
> Non-iframe

Should this be listed as "Option 0: Non-iframe", so it can be clearer what the "Pro" and "Con" sections are talking about below?

Reply:
I am not sure what you mean here. Non-iframe is a class of solution which has two specific solutions under `option 1` and `option 2`.